src/module_library/root_onedim.h

```diff
@@ -0,0 +1,697 @@
+#ifndef ROOTS_ONEDIM_H
+#define ROOTS_ONEDIM_H
+
+#include <cmath>
+#include <limits>
+#include <string>
+
+namespace root_algorithm
+{
+
+// Helper function declarations
+inline bool is_close(double x, double y, double tol, double rtol);
+inline bool is_close(double x, double y);
+inline bool is_zero(double x, double tol);
+inline bool is_zero(double x);
+inline bool same_signs(double x, double y);
+inline bool opposite_signs(double x, double y);
+inline bool smaller(double x, double y);
+inline bool is_between(double x, double a, double b);
+
+// For error handling.
+enum class Flag {
+    valid,  // don't terminate
+    residual_zero,
+    delta_root_zero,
+    bracket_width_zero,
+    max_iterations,
+    invalid_bracket,
+    division_by_zero,
+    halley_no_cross
+};
+inline bool successful_termination(Flag flag);
+inline std::string flag_message(Flag flag);
+
+/*
+For error handling, this struct adds flags to a state used in the
+iterations. This is designed to be a monad, meaning the following rules
+should be respected: For any function/function object with the signature:
+`T f(T x)` there is a function `with_flag<T> f(with_flag<T> x)` such that
+if (is_valid(x)) return with_flag<T>{f(x.state)}; else return x;
+
+*/
+template <typename T>
+struct with_flag {
+    T state;
+    Flag flag;
+
+    with_flag() = default;
+    with_flag(T x) : state(x), flag(Flag::valid) {}
+    with_flag(T x, Flag f) : state(x), flag(f) {}
+    with_flag(Flag f) : state{}, flag(f) {}
+};
+
+template <typename T>
+inline bool is_valid(with_flag<T> x)
+{
+    return x.flag == Flag::valid;
+}
+
+/**
+ * @class result_t
+ *
+ * @brief Result from root finding algorithm.
+ *
+ * @param root The identified root.
+ *
+ * @param residual The output value (root); should equal zero if
+ * successful.
+ *
+ * @param iteration The number of iterations performed.
+ *
+ * @param flag Indicates the reason for termination.
+ *
+ */
+struct result_t {
+    double root;
+    double residual;
+    size_t iteration;
+    Flag flag;
+};
+
+/**
+ * @class root_finder
+ *
+ * @brief Function object for finding the root or zero of a function.
+ *
+ * @param [in] max_iter The total number of iterations allowed.
+ *
+ * @param [in] abs_tol The absolute tolerance, the error threshold for
+ * floating-point zero. E.g., abs_tol = 1e-12 means |x| < 1e-12, or x equals
+ * zero to 12 digits. Used for convergence testing.
+ *
+ * @param [in] rel_tol The relative tolerance, the error threshold for
+ * comparing floating point number equality. E.g., rel_tol = 1e-12 means
+ * x equals y if the first 12 digits match. Used for convergernce testing.
+ *
+ * @details This class creates a function object whose solve or call method
+ * is an interface for using any of the root finding methods. A class template
```

Comment:
Once an object of this class has been created, it is an interface for one of the root finding methods, not any of the root finding methods.

I would suggest rewording this to something like

"This class standardizes the interface to each of the root finding methods."

Or

"When instantiated, this class is a function object whose solve method is an interface to one of the root finding methods."